trio/_socket.py

```diff
@@ -451,36 +548,69 @@ def __init__(self, sock: _stdlib_socket.socket):
     # Simple + portable methods and attributes
     ################################################################
 
-    # NB this doesn't work because for loops don't create a scope
-    # for _name in [
-    #         ]:
-    #     _meth = getattr(_stdlib_socket.socket, _name)
-    #     @_wraps(_meth, assigned=("__name__", "__doc__"), updated=())
-    #     def _wrapped(self, *args, **kwargs):
-    #         return getattr(self._sock, _meth)(*args, **kwargs)
-    #     locals()[_meth] = _wrapped
-    # del _name, _meth, _wrapped
-
-    _forward = {
-        "detach",
-        "get_inheritable",
-        "set_inheritable",
-        "fileno",
-        "getpeername",
-        "getsockname",
-        "getsockopt",
-        "setsockopt",
-        "listen",
-        "share",
-    }
-
-    def __getattr__(self, name):
-        if name in self._forward:
-            return getattr(self._sock, name)
-        raise AttributeError(name)
-
-    def __dir__(self) -> Iterable[str]:
-        return [*super().__dir__(), *self._forward]
+    # forwarded methods
+    def detach(self) -> int:
+        return self._sock.detach()
+
+    def fileno(self) -> int:
+        return self._sock.fileno()
+
+    def getpeername(self) -> Any:
+        return self._sock.getpeername()
+
+    def getsockname(self) -> Any:
+        return self._sock.getsockname()
+
+    @overload
+    def getsockopt(self, /, level: int, optname: int) -> int:
+        ...
+
+    @overload
+    def getsockopt(self, /, level: int, optname: int, buflen: int) -> bytes:
+        ...
+
+    def getsockopt(
+        self, /, level: int, optname: int, buflen: int | None = None
+    ) -> int | bytes:
+        if buflen is None:
+            return self._sock.getsockopt(level, optname)
+        return self._sock.getsockopt(level, optname, buflen)
+
+    @overload
+    def setsockopt(self, /, level: int, optname: int, value: int | Buffer) -> None:
+        ...
+
+    @overload
+    def setsockopt(self, /, level: int, optname: int, value: None, optlen: int) -> None:
+        ...
+
+    def setsockopt(
+        self,
+        /,
+        level: int,
+        optname: int,
+        value: int | Buffer | None,
+        optlen: int | None = None,
+    ) -> None:
+        if optlen is None:
+            return self._sock.setsockopt(level, optname, cast("int|Buffer", value))
+        return self._sock.setsockopt(level, optname, cast(None, value), optlen)
```

Comment:
Looks like we need a test using `value = None`. Instead of the casts here, perhaps it’d better to add an a check in each side of the branch to verify that value is or is not `None`, so that can be eliminated from the union. Better to be a `raise` instead of `assert` since it’s user code that would be wrong.

Reply:
Implemented :+1:

Reply:
The first exception message is wrong, `optlen` isn’t set there?